### DISCUSSION TITLE
Suggest installing @types/scope__name when users try to install @types/@scope/name

### DISCUSSION THREAD

**Discussion by jeswr** [npm/feedback#608] [CLI] [answered]
When installing typescript declarations for packages, they usually follow the form `@types/package_name` - this means that devs may try to install `@types/@scope/name` when installing scoped packages (I did this a lot when I started out and still do it from time to time when I am tired).

Since the DefinitelyTyped convention is to package scoped packages as `@types/scope__name` instead It would be good to have a CLI hint saying something like

```
You tried to install @types/@scope/name which is not a valid package name.
Would you like to install @types/scope__name instead (Y/N)?
```

instead of immediately throwing the error

```console
npm ERR! code EINVALIDPACKAGENAME
npm ERR! Invalid package name "@types/": name can only contain URL-friendly characters

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jeswr/.npm/_logs/2021-12-03T09_50_07_680Z-debug.log
```

**Comment by ljharb** [comment]
What version would it install? DT versions aren’t 1:1 and sometimes don’t follow the major, let alone the minor version of a package.

**Reply by jeswr** [reply]
If the version is not specified then I think it should be the same as running `npm install @types/scope__name` - which AFAIK would install the latest version, and if the user tries to specify a version; for instance `npm install @types/@scope/name@version` then the version should be referencing the DT version number since this is what happens in the case of installing a package of the form `npm install @types/package-name@version`

In order to clarify a bit further I have included some examples of what I think the behavior should be below:


Version not specified
```
$ npm install @types/@scope/name
You tried to install @types/@scope/name which is not a valid package name.
Would you like to install @types/scope__name instead (Y/N)? Y
> npm install @types/scope__name
```

Version specified
```
$ npm install @types/@scope/name@version
You tried to install @types/@scope/name which is not a valid package name.
Would you like to install @types/scope__name instead (Y/N)? Y
> npm install @types/scope__name@version
```

Flags included
```
$ npm install -D @types/@scope/name
You tried to install @types/@scope/name which is not a valid package name.
Would you like to install @types/scope__name instead (Y/N)? Y
> npm install -D @types/scope__name
```
Multiple packages
```
$ npm install @types/@scope/name @types/@scope2/name2 @types/unscoped-package

You tried to install @types/@scope/name which is not a valid package name.
Would you like to install @types/scope__name instead (Y/N)? Y

You tried to install @types/@scope2/name2 which is not a valid package name.
Would you like to install @types/scope2__name2 instead (Y/N)? Y

> npm install @types/scope__name @types/scope2__name2 @types/unscoped-package
```

Selecting `N`
```
$ npm install -D @types/@scope/name@version
You tried to install @types/@scope/name which is not a valid package name.
Would you like to install @types/scope__name instead (Y/N)? N

npm ERR! code EINVALIDPACKAGENAME
npm ERR! Invalid package name "@types/": name can only contain URL-friendly characters

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jeswr/.npm/_logs/2021-12-03T09_50_07_680Z-debug.log
```

**Comment by MylesBorins** [comment, marked as answer]
There isn't historic precedence in the npm CLI for having error messages that are specific to a certain communities patterns, so I think it would be a hard sell to adopt this change.

With that said, if you really want to pursue this change the next step would be to kick off an rfc in the [npm rfc repo](https://github.com/npm/rfcs)
